File: src/java/WebNinja/src/test/java/webninja/selenuim/CodeTrack.java
package webninja.selenuim;

import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.support.ui.Select;
import webninja.UserRepository;

public class CodeTrack {
    private final String baseUrl;
    private final WebDriver webDriver;
    private final UserRepository userRepository;

    public CodeTrack(
            String baseUrl,
            WebDriver webDriver,
            UserRepository userRepository) {
        this.baseUrl = baseUrl;
        this.webDriver = webDriver;
        this.userRepository = userRepository;
    }



    public void logOut() {
        webDriver.findElement(
                By.xpath("//a[@title='Log off the CodeTrack System']"))
                .click();
    }



    public LogonPage gotoLogonPage() {
        webDriver.get(getUrl("?page=login"));
        return new LogonPage(webDriver);
    }



    public HomePage gotoHomePage() {
        webDriver.findElement(
                By.xpath("//a[@title='Summary of your current project']"))
                .click();
        return new HomePage(webDriver, this, userRepository);
    }



    public boolean isLoggedIn() {
        webDriver.get(getUrl(""));
        try {
            gotoHomePage();
            return true;
        }
        catch (NoSuchElementException e) {
            return false;
        }
    }



    public AdminPage showAdminPage() {
        webDriver.findElement(
                By.xpath("//a[@title='CodeTrack Administration and Setup']"))
                .click();
        return new AdminPage(webDriver);
    }



    public IssueForm showNewIssueForm() {
        webDriver.findElement(
                By.xpath("//a[@title='Create a new defect report or Change Request']"))
                .click();
        return new IssueForm(webDriver, this, userRepository);
    }




    public ReportsPage gotoReportsPage() {
        webDriver.findElement(
                By.xpath("//a[@title='Create simple and advanced reports']"))
                .click();
        return new ReportsPage(webDriver);
    }



    public void selectOption(By by, String text) {
        Select select = new Select(webDriver.findElement(by));
        select.selectByVisibleText(text);
    }



    public void setProjectName(String projectName) {
        selectOption(By.name("project"), projectName);
    }

    public void submit() {
        webDriver.findElement(
                By.xpath("//input[@type='submit']"))
                .click();
    }



    private String getUrl(String relativePath) {
        return String.format("%s%s", baseUrl, relativePath);
    }

}

File: src/dotnet/WebNinja/WebNinja/selenium/CodeTrack.cs
using System;
using OpenQA.Selenium;
using WebNinja.webninja;

namespace WebNinja.selenium
{
    public class CodeTrack
    {
        private readonly string _baseUrl;
        private readonly IWebDriver _driver;
        private readonly UserRepository _userRepository;

        public CodeTrack(string baseUrl, IWebDriver driver, UserRepository userRepository)
        {
            _baseUrl = baseUrl;
            _driver = driver;
            _userRepository = userRepository;
        }

        public bool IsLoggedIn
        {
            get
            {
                _driver.Url = Url("");
                try
                {
                    GotoHomePage();
                    return true;
                }
                catch (NoSuchElementException)
                {
                    return false;
                }
            }
        }

        public LogonPage GotoLogonPage()
        {
            _driver.Url = Url("?page=login");
            return new LogonPage(_driver);
        }

        public AdminPage GotoAdminPage()
        {
            _driver.FindElement(By.XPath("//a[@title='CodeTrack Administration and Setup']")).Click();
            return new AdminPage(_driver);
        }

        public HomePage GotoHomePage()
        {
            _driver.FindElement(By.XPath("//a[@title='Summary of your current project']")).Click();
            return new HomePage(_driver, _userRepository);
        }

        public void Logout()
        {
            _driver.FindElement(By.XPath("//a[@title='Log off the CodeTrack System']")).Click();
        }

        public IssueForm GotoNewIssueForm()
        {
            _driver.FindElement(By.XPath("//a[@title='Create a new defect report or Change Request']")).Click();
            return new IssueForm(_driver, _userRepository);
        }

        public ReportsPage GotoReportsPage()
        {
            _driver.FindElement(By.XPath("//a[@title='Create simple and advanced reports']")).Click();
            return new ReportsPage(_driver);
        }

        private String Url(String relativePath)
        {
            return String.Format("{0}{1}", _baseUrl, relativePath);
        }

    }
}
